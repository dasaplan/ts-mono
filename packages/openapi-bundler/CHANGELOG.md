# @dasaplan/openapi-bundler

## 0.0.7

### Patch Changes

- ef2415e: fixed toposort schemas when array items are allOf objects

## 0.0.6

### Patch Changes

- fix mergeAlOf order to follow default behaviour

## 0.0.5

### Patch Changes

- fix mergeAllOf order to follow default behaviour

## 0.0.3

### Patch Changes

- Update documentation

## 0.0.2

### Patch Changes

- handle dangling properties for schemas with allOf array.
