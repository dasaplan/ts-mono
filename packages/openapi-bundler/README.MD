# @dasaplan/openapi-bundler

- âœ… Configured wrapper of the standard [redocly](https://redocly.com/docs/cli/) bundler.
- âœ… Includes opinionated post-processing steps to transform the spec in a OpenApi 3.0.3 standard compliant way (see [spec-rules][spec-rules]).
- âœ… Is implemented against OpenApi 3.0.3.

## Getting started
### Prerequisite
- using npm
    ```shell
    # integrate in project
    npm i --save-dev @dasaplan/openapi-bundler
    ```    
    ```shell
    # global installation
    npm i -g @dasaplan/openapi-bundler
    ```    
- using pnpm
    ```shell
    # integrate in project
    pnpm i --save-dev @dasaplan/openapi-bundler
    ```    
    ```shell
    # global installation
    pnpm i -g @dasaplan/openapi-bundler
    ```    

### usage
- Bundle an OpenapiSpec
  ```shell
  # assuming the root spec file is located at "$cwd/specs/generic/api.yml"
  # and assuming we want all generated files to find at "$cwd/out"
  openapi-bundler-cli bundle specs/generic/api.yml -o ./out/bundled-api.yml
  ```
- See help for more options
    ```shell
    openapi-bundler-cli bundle --help
    ```
  
## Features
- âœ… Configured wrapper of the standard [redocly](https://redocly.com/docs/cli/) bundler.
> note: Example specs are written in a compressed but valid way to reduce whitespace and facilitate an overview over the whole example.

### post processing
<details>
<summary>âœ… merge allOf subschemas</summary>

<details style="margin-left: 20px"><summary> motivation </summary>
Merging allOf arrays means we reduce complexity for generator and documentation tooling in order to facilitate consistency across tech stacks. 
<p>The easiest way would be to flat out every allOf so there resulting spec does not have the complexity at all. Yet, we need to find a middle ground because codegenerators may rely on identifiers like the discriminator to create object hierarchies.</p> 
We could just let every tool care for themselves how to handle their use-cases. However, from experience this will yield undefined behaviour for API consumer and lead to frustration because the default tooling does no yield results aligning with the specification.

```yaml
# schema declares $ref property but also other properties:
# most tools ignore everything besides $ref
# -> ðŸ”¥ unspecified behavior across tooling and use cases
RefSchema_with_dangling_properties:
  $ref: '#/A_Schema'
  description: "This is something I like to share!"
  title: RefSchema_with_dangling_properties

# schema declares allOf but declares further properties:
# most tool will merge siblings into array, but some may ignore them 
# -> ðŸ”¥ inconsistent behavior across tooling and use cases
AllOfSchema_with_dangling_properties:
  allOf: [ { $ref: '#/A_Schema'} ]
  properties: # sibling to allOf should be part of allOF
    name: { type: string }

# schema declares allOf with multiple elements:
# no problem for validators but code gen mostly try to merge them
# -> ðŸ”¥ merge behaviour may yield unexpected results
AllOfSchema_with_multi_elements:
  allOf:
    - $ref: '#/A_Schema',
    - $ref: '#/B_Schema',
    - properties: { foo: { type: string } }
```
</details>


<details style="margin-left: 20px">
<summary> src spec </summary>

````yaml
components:
  schemas:
    PetBase: 
      type: object,
      properties: { type: { type: string } }
    
    CatBase:  
      type: object,
      discriminator: { propertyName: catType },
      properties: 
        type: { type: string }, 
        catType: { type: string, enum: [ 'SEAM', 'SHORT_HAIR' ] } 

    SeamCat:
      allOf:
        - $ref: '#/components/PetBase'
        - $ref: '#/components/CatBase'
        - title: SeamCat
    ShortHairCat:
      allOf:
        - $ref: '#/components/PetBase'
        - $ref: '#/components/CatBase'
        - title: ShortHairCat

    Dog: { allOf: [ { $ref: '#/components/PetBase' }, { title: Dog } ] }



````
</details>

<details style="margin-left: 20px">
<summary> resulting spec </summary>

````yaml
components:
  schemas:
    PetBase: { type: object, required: [type] discriminator: {propertyName: type},properties: { type: { type: string } } }

    SeamCat:
      allOf:
        - $ref: '#/components/PetBase'
        - { title: SeamCat, type: object, properties: { type: { catType: string, enum: [ 'SEAM', 'SHORT_HAIR' ] } } }

    ShortHairCat:
      allOf:
        - $ref: '#/components/PetBase'
        - { title: ShortHairCat, type: object, properties: { type: { catType: string, enum: [ 'SEAM', 'SHORT_HAIR' ] } } }

    Dog:
      allOf:
        - $ref: '#/components/PetBase'
        - title: Dog 
````
</details>
</details>



<details>
<summary>âœ… ensure discriminator values</summary>

- ensures that:
    - discriminator property type is of type string, and not enum.
        - most sophisticated tools can infer the values from the explicit or implicit discriminator mapping
    - polymorphism and inheritance can be explicitly inferred from the spec
        - every polymorph subschema defines the discriminator property and respective value
            - this is only necessary to extend the standard generator with templating
            - discriminator value is declared with <b>x-const</b> to avoid triggering compatibility layers for discriminator
        - any parent schema referenced from an allOf array does not define discriminator mapping
        - discriminator mapping only exists on schemas with a oneOf member

</details>
<details>
<summary>âœ… support multi schema usage in discriminator mapping (1:n discriminator value to mapping)</summary>

````yaml
Cat:
  type: object
  properties:
    type: 
      type: string
      x-const: [ 'SEAM', 'SHORT_HAIR' ]
    
MyOneOfSchema:
  oneOf: [{$ref: '#/Dog', {$ref: '#/Cat'}}]
  discriminator:
    propertyName: 'type'
    mapping:
      SEAM: "#/Cat"
      SHORT_HAIR: "#/Cat"
      Dog: "#/Dog"
````
</details>
<p></p>
<details>
<summary> <b style="color: #2c6fdf">ðŸ“œ complete example: openapi post-processing </b> </summary>

| examples | src                                 | bundled                                     | post-processed                                       |
|----------|-------------------------------------|---------------------------------------------|------------------------------------------------------|
| simple   | [link to src][simple-petstore-src]  | [link to bundled][simple-petstore-bundled]  | [link to post-processed][simple-petstore-processed]  |
| complex  | [link to src][complex-petstore-src] | [link to bundled][complex-petstore-bundled] | [link to post-processed][complex-petstore-processed] |

</details>

[complex-petstore-src]: ../openapi-specs/specs/pets-modular-complex/petstore-api.yml

[complex-petstore-bundled]: ./docs/examples/specs/complex-petstore-bundled.yml

[complex-petstore-processed]: ./docs/examples/specs/complex-petstore-post-processed.yml

[simple-petstore-src]: ../openapi-specs/specs/pets-simple/pets-api.yml

[simple-petstore-bundled]: ./docs/examples/specs/simple-petstore-bundled.yml

[simple-petstore-processed]: ./docs/examples/specs/simple-petstore-post-processed.yml

[spec-rules]: ./docs/spec-rules.md