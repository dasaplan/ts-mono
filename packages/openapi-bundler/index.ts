export {
  bundleOpenapi,
  createCommandBundle,
  createSpecProcessor,
  cleanObj,
  OpenApiBundled,
  parseOpenapi,
  Resolver,
  Transpiler,
  OaComponent,
  SchemaResolverContext,
  Schema,
} from "./src/index.js";
