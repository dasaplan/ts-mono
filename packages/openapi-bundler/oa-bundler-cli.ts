#!/usr/bin/env ts-node

import { Command } from "commander";
import * as process from "process";
import { createCommandBundle } from "./src/index.js";

const program = new Command();
createCommandBundle(program);

program.parse(process.argv);
