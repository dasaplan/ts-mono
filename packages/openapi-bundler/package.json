{
  "name": "@dasaplan/openapi-bundler",
  "description": "Opinionated bundler/merger for Openapi (Swagger)",
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "type": "module",
  "version": "0.0.0",
  "keywords": [
    "openapi",
    "typescript",
    "ts",
    "zod",
    "codegen"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/dasaplan/ts-mono.git",
    "directory": "packages/openapi-bundler"
  },
  "scripts": {
    "prepack": "sort-package-json --quiet && clean-package",
    "postpack": "clean-package restore",
    "build": "pnpm rup",
    "rup": "tsup-node --dts",
    "lint": "eslint \"./src/*.ts\" index.ts oa-bundler-cli.ts",
    "clean": "rimraf dist",
    "test": "pnpm test:esm",
    "test:esm": "cross-env NODE_OPTIONS=--experimental-vm-modules jest",
    "test:esm:update": "cross-env NODE_OPTIONS=--experimental-vm-modules jest -u"
  },
  "bin": {
    "openapi-bundler-cli": "dist/oa-cli.js"
  },
  "files": ["dist"],
  "dependencies": {
    "commander": "^11.1.0",
    "json-schema-merge-allof": "^0.8.1",
    "jsonpointer": "^5.0.1",
    "openapi3-ts": "^4.3.3",
    "merge-json-schemas": "^1.0.0",
    "tslog": "^4.9.2",
    "lodash": "^4.17.21",
    "toposort": "^2.0.2"
  },
  "peerDependencies": {
    "@redocly/openapi-core": "^1.6.0"
  },
  "devDependencies": {
    "tsconfig": "workspace:*",
    "@types/json-schema-merge-allof": "^0.6.5",
    "@types/lodash":"^4.17.5",
    "@dasaplan/ts-sdk": "workspace:*"
  },
  "clean-package": {
    "remove": [
      "scripts",
      "devDependencies",
      "tsup",
      "clean-package",
      "dependencies.@dasaplan/ts-sdk"
    ]
  },
  "tsup": {
    "entry": [
      "index.ts",
      "oa-bundler-cli.ts"
    ],
    "noExternal": [
      "@dasaplan/ts-sdk",
      "tsconfig"
    ],
    "format": [
      "cjs",
      "esm"
    ],
    "tsconfig": "tsconfig.build.json",
    "clean": true,
    "sourcemap": true
  }
}
