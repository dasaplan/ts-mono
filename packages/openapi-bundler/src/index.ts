export { bundleOpenapi, OpenApiBundled, parseOpenapi } from "./bundle.js";
export { createCommandBundle } from "./commands.js";
export { createSpecProcessor } from "./post-process/index.js";
export { Transpiler, Schema } from "./transpiler/index.js";
export {
  Resolver,
  OaComponent,
  SchemaResolverContext,
  cleanObj,
} from "./resolver/index.js";
