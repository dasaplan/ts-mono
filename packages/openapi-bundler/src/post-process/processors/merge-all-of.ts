/* eslint-disable @typescript-eslint/no-explicit-any,@typescript-eslint/no-unused-vars */

import { OpenApiBundled } from "../../bundle.js";
import { oas30 } from "openapi3-ts";

import jsonSchemaMergeAllOff from "json-schema-merge-allof";
import { isRef } from "@redocly/openapi-core";
import { appLog } from "../../logger.js";
import { _, ApplicationError } from "@dasaplan/ts-sdk";
import { cleanObj, SchemaResolverContext } from "../../resolver/index.js";
import { mergeXOmit, XOmitConfig } from "./x-omit-deep.js";

export function mergeAllOf(bundled: OpenApiBundled) {
  const log = appLog.childLog(mergeAllOf);

  const mergedAllOf = _.cloneDeep(bundled);
  const { collected, ctx } = findSchemaObjectsWithAllOf(mergedAllOf);
  log.info(`mergeAllOf. Merging allOf arrays in ${collected.length} schemas`);
  log.debug(`mergeAllOf. Merging in schemas: ${collected.map((s) => s.id).join(", ")}`);

  collected.forEach((s) => {
    try {
      doMerge(s, ctx);
    } catch (e) {
      throw ApplicationError.create(`failed to merge allOf for schema.id ${s.id}`).chainUnknown(e);
    }
  });
  return mergedAllOf;
}

function doMerge({ schema, id }: { id: string; schema: any }, ctx: SchemaResolverContext) {
  const log = appLog.childLog(doMerge);
  const subSchemas: Array<oas30.ReferenceObject | oas30.SchemaObject> = schema.allOf ?? [];
  // include dangling properties {allOf: [], danglingA: {}, danglingB: [], danglingC: null, ... }
  const danglingProperties = { ..._.omit(schema, "allOf") };
  if (!_.isEmpty(danglingProperties)) {
    log.debug(`danglingProperties. fixing dangling properties for schema.id ${id} - moving all properties into allOf array`);
    schema.allOf?.push(danglingProperties);
    cleanObj(schema, ["allOf"]);
  }
  // when we merge allOf array, we do not want to mutate referenced entities
  // omitting clone will mutate "schema"
  const clonedSchemas = _.cloneDeep(subSchemas);
  // omitting clone will mutate cached references
  const resolvedSchemas = _.cloneDeep(resolveSubSchemas(clonedSchemas, ctx));
  if (!_.isEmpty(danglingProperties)) {
    resolvedSchemas.push({ pointer: undefined, resolved: danglingProperties });
  }
  const hasOneOfOrAnyOf = resolvedSchemas.filter((f) => !_.isEmpty(f.resolved.oneOf) || !_.isEmpty(f.resolved.anyOf))?.length > 0;
  if (hasOneOfOrAnyOf) {
    throw ApplicationError.create(`unsupported allOf element. anyOf or oneOf are yet not supported as an allOf subschema: ${JSON.stringify(schema)}`);
  }

  // scenario(real inheritance): a allOf.element holds a ref where the target has a discriminator
  const parentsWithDiscriminator = resolvedSchemas.filter((p) => _.isDefined(p.pointer) && _.isDefined(p.resolved.discriminator));

  // scenario(composition): a allOf.element is a schema
  const resolvedSchemasWithoutDiscriminator = resolvedSchemas.filter((p) => !_.isDefined(p.resolved.discriminator));

  // scenario(composition of a base schema): a allOf.element defines a schema with a discriminator by reference and inlined
  const inlinesWithDiscriminator = resolvedSchemas.filter((p) => !_.isDefined(p.pointer) && _.isDefined(p.resolved.discriminator));
  if (parentsWithDiscriminator.length == 1 && inlinesWithDiscriminator.length > 0) {
    // we may have an allOf expressing inheritance and multiple schemas we need to merge (while keeping inherited schema)
    const mergedWithInlineParent = mergeSubSchemas([...inlinesWithDiscriminator, ...resolvedSchemasWithoutDiscriminator], ctx);
    // build hierarchy
    schema.allOf = [{ $ref: parentsWithDiscriminator[0].pointer }, mergedWithInlineParent.resolved];
    return schema;
  }
  if (resolvedSchemas.length < 1) {
    // something is off: could be a schema with a single allOf or an allOf comprised of multiple discriminators...
    log.warn(`found allOf with single element in schema.id '${id}': ${JSON.stringify(schema)}`);
    return schema;
  }

  if (parentsWithDiscriminator.length == 1 && resolvedSchemasWithoutDiscriminator.length == 1) {
    // we may have an allOf expressing inheritance - most tooling can handle two allOfs
    return schema;
  }

  if (parentsWithDiscriminator.length == 1 && resolvedSchemasWithoutDiscriminator.length > 1) {
    // we may have an allOf expressing inheritance and multiple schemas we need to merge (while keeping inherited schema)
    const merged = mergeSubSchemas(resolvedSchemasWithoutDiscriminator, ctx);
    // build hierarchy
    schema.allOf = [{ $ref: parentsWithDiscriminator[0].pointer }, merged.resolved];
    return schema;
  }

  if (parentsWithDiscriminator.length > 1) {
    // we may have an allOf expressing multiple inheritance we will try to find a common base

    const merged = mergeSubSchemas(resolvedSchemasWithoutDiscriminator, ctx);
    // build hierarchy
    const hierarchy = parentsWithDiscriminator.reduce((acc, curr) => {
      return { allOf: [{ $ref: curr.pointer }, acc] };
    }, merged.resolved);
    delete schema["$ref"];
    Object.assign(schema, hierarchy);
    return schema;
  }

  if (parentsWithDiscriminator.length < 1) {
    // we only have schemas which we can merge together
    const merged = mergeSubSchemas(resolvedSchemasWithoutDiscriminator, ctx);
    Object.assign(schema, merged.resolved);
    delete schema["allOf"];
    delete schema["$ref"];
    return schema;
  }

  throw "Error: should not have been reached";
}

function resolveSubSchemas(
  subSchemas: Array<oas30.ReferenceObject | oas30.SchemaObject>,
  ctx: SchemaResolverContext
): Array<{
  pointer: string | undefined;
  resolved: oas30.SchemaObject & { $ref?: string };
}> {
  const subs = subSchemas.map((d) => {
    return resolveRefNode(d, ctx, { deleteRef: true });
  });
  const allOfs = subs.flatMap((s) => s.resolved.allOf ?? []);
  if (_.isEmpty(allOfs)) {
    return subs;
  }
  const flatten = resolveSubSchemas(allOfs, ctx);
  return [...flatten, ...subs].map((s) => ({
    pointer: s.pointer,
    resolved: _.omit(s.resolved, "allOf"),
  }));
}

function getJsonSchemaMergeAllOff<T extends oas30.SchemaObject>(
  subschemas: Array<{
    pointer: string | undefined;
    resolved: T;
  }>
): T {
  try {
    return jsonSchemaMergeAllOff<oas30.SchemaObject & { "x-omit"?: XOmitConfig }>(
      { allOf: subschemas.map((s) => s.resolved) },
      {
        resolvers: {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-expect-error
          additionalProperties: (values, path, mergeSchemas, options) => {
            return jsonSchemaMergeAllOff.options.resolvers.additionalProperties(values, path, mergeSchemas, options);
          },
          // overwrite title by higher indexed schema: [ {title: a}, {title:b}] => {title:b}
          // default title means take first occurance
          defaultResolver: (values, path, mergeSchemas, options) => {
            if (path?.[0] === "x-omit") {
              const [a, b] = values;
              return mergeXOmit(a, b);
            }

            if (options.resolvers?.title) {
              // use configured resolver
              return options.resolvers.title(values, path, mergeSchemas, options);
            }
            // use fallback resolver
            return jsonSchemaMergeAllOff.options.resolvers.title(values, path, mergeSchemas, options);
          },
        },
      }
    ) as T;
  } catch (e: unknown) {
    throw ApplicationError.create("failed merging allOf sub schemas").chainUnknown(e);
  }
}

export function mergeSubSchemas(
  _resolvedSchemas: Array<{
    pointer: string | undefined;
    resolved: oas30.SchemaObject;
  }>,
  ctx: SchemaResolverContext
) {
  const subschemas = _resolvedSchemas.map((d) => ({
    pointer: d.pointer,
    resolved: ctx.resolver.resolveRefImmutable(d.resolved, { deleteRef: true }),
  }));

  const merged = getJsonSchemaMergeAllOff(subschemas.reverse());
  // clean references in properties
  Object.values(merged.properties ?? {}).forEach((propValue) => {
    if (!isRef(propValue) || _.isEmpty(_.omit(propValue, "$ref"))) {
      // property is either a ref or an object => that is fine
      return;
    }
    // we have both a ref and an object. We try to keep the ref, if the content is identical
    const schema = {
      pointer: undefined,
      resolved: _.cloneDeep(_.omit(propValue, "$ref")),
    };
    const referenced = resolveRefNode(propValue, ctx, { deleteRef: false });
    if (_.isEqual(schema.resolved, referenced.resolved)) {
      cleanObj(propValue, []);
      Object.assign(propValue, { $ref: referenced.pointer });
      return;
    }
    Object.assign(propValue, getJsonSchemaMergeAllOff([schema, referenced]));
    delete (propValue as never)["$ref"];
  });
  return {
    pointer: undefined,
    resolved: merged,
  };
}

function findSchemaObjectsWithAllOf(bundled: OpenApiBundled) {
  const resolver = SchemaResolverContext.create(bundled);
  const collected = resolver.schemas.filter((s) => _.isDefined(s.schema.allOf));
  return { collected, ctx: resolver };
}

function resolveRefNode(data: { $ref: string } | unknown, ctx: SchemaResolverContext, params?: { deleteRef: boolean }) {
  if (!isRef(data)) {
    return {
      pointer: undefined,
      resolved: ctx.resolver.resolveRefImmutable(data as oas30.SchemaObject, params),
    };
  }
  return {
    pointer: data["$ref"],
    resolved: ctx.resolver.resolveRefImmutable(data, params),
  };
}
