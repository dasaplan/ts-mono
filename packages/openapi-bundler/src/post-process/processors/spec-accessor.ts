import { OpenApiBundled } from "../../bundle.js";
import { Resolver } from "../../resolver/index.js";
import { ApplicationError } from "@dasaplan/ts-sdk";

export namespace OpenapiApiDoc {
  type ApiDoc = OpenApiBundled;

  export function accessor(bundled: ApiDoc) {
    const resolver = Resolver.create(bundled);
    return {
      getSchemaByName: createGetSchemaByName(bundled, resolver),
      getSchemaByRef: createGetSchemaByRef(bundled, resolver),
      get schemas() {
        return Object.entries(bundled.components?.schemas ?? {}).map(([name, schema]) => ({ name, schema }));
      },
    };
  }

  function createGetSchemaByName(oa: ApiDoc, resolver: Resolver) {
    return (name: string) => {
      const schema = oa.components?.schemas?.[name];
      ApplicationError.assert(schema, `could not find schema.name in spec: ${name} `);
      return resolver.resolveRef(schema, { deleteRef: true });
    };
  }

  function createGetSchemaByRef(oa: ApiDoc, resolver: Resolver) {
    return (ref: string | { $ref: string }) => {
      const refObj = typeof ref === "string" ? { $ref: ref } : ref;
      const schema = resolver.resolveRef(refObj);
      ApplicationError.assert(schema, `could not find schema.$ref in spec`);
      return schema;
    };
  }
}
