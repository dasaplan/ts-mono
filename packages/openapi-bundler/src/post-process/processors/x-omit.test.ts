/* eslint-disable @typescript-eslint/no-unused-vars */
import { describe, expect, test } from "vitest";
import { mergeAllOf } from "./merge-all-of.js";
import { BundleMock } from "@dasaplan/openapi-bundler";

import { OpenapiApiDoc } from "./spec-accessor.js";

describe("x-omit", () => {
  const {
    createApi,
    withSchema,
    factory: { schemaRef, mockXOmit },
  } = BundleMock.create();

  test("omits", () => {
    const spec = createApi(
      withSchema("A", {
        required: ["a"],
        properties: { a: { type: "string" } },
      }),
      withSchema("B", {
        allOf: [
          { required: ["b", "bb"], properties: { b: { type: "string" }, bb: { type: "string" } } },
          mockXOmit({
            required: ["bb"],
            properties: { bb: true },
          }),
        ],
      }),
      withSchema("AB", {
        allOf: [
          schemaRef("A"),
          schemaRef("B"),
          mockXOmit({
            required: ["a"],
            properties: { b: true },
          }),
        ],
      })
    );

    const merged = mergeAllOf(spec);
    expect(merged).toMatchSnapshot();
    const accessor = OpenapiApiDoc.accessor(merged);
    expect(accessor.getSchemaByName("A"), "expected schema 'A' to be defined").toBeDefined();
    expect(accessor.getSchemaByName("B"), "expected schema 'b' to be defined").toBeDefined();

    const AB_afterMerge = accessor.getSchemaByName("AB");
    expect(AB_afterMerge.required, "expected 'required' not to be processed").toEqual(["a", "b", "bb"]);
    expect(AB_afterMerge["x-omit"], "expected x-omit to be defined").toBeDefined();
    expect(AB_afterMerge.properties?.a, "expected property 'a' not to be processed").toBeDefined();
    expect(AB_afterMerge.properties?.a, "expected property 'b' not to be processed").toBeDefined();

    // x-omit
    // todo
    // expect(AbfterMerge["x-omit"]).toBeUndefined();
    // expect(AbfterMerge.allOf).toBeUndefined();
    // expect(AbfterMerge.required).toEqual(["b"]);
    // expect(AbfterMerge.properties?.a).toBeDefined();
    // expect(AbfterMerge.properties?.b).toBeUndefined();
    // expect(merged).toMatchSnapshot();
  });
});
