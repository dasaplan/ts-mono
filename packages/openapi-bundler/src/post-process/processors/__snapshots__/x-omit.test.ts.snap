// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`x-omit > omits 1`] = `
{
  "components": {
    "schemas": {
      "A": {
        "properties": {
          "a": {
            "type": "string",
          },
        },
        "required": [
          "a",
        ],
        "type": "object",
      },
      "AB": {
        "properties": {
          "a": {
            "type": "string",
          },
          "b": {
            "type": "string",
          },
          "bb": {
            "type": "string",
          },
        },
        "required": [
          "a",
          "b",
          "bb",
        ],
        "type": "object",
        "x-omit": {
          "properties": {
            "b": true,
            "bb": true,
          },
          "required": [
            "a",
            "bb",
          ],
        },
      },
      "B": {
        "properties": {
          "b": {
            "type": "string",
          },
          "bb": {
            "type": "string",
          },
        },
        "required": [
          "b",
          "bb",
        ],
        "x-omit": {
          "properties": {
            "bb": true,
          },
          "required": [
            "bb",
          ],
        },
      },
    },
  },
  "info": {
    "title": "",
    "version": "",
  },
  "openapi": "3.0.3",
  "paths": {},
}
`;
